From 7cf3677adb2df96e848b66bfb6aea674bee75f21 Mon Sep 17 00:00:00 2001
From: Ferry Toth <ftoth@exalondelft.nl>
Date: Sat, 14 Jan 2023 23:10:33 +0100
Subject: [PATCH 1/1] Add shared ninja jobserver support

This patch depends on the patch "Add shared make jobserver support"
for creating a named pipe with execution tokens.
This allows both make and ninja to share the same pool of tokens allowing
the total number of compile tasks to be restricted to the number of threads.
On kirkstone cmake jobs rely on ninja for building so those benefit as well.

The patch works by fetching the ninja clone with patches to support
jobserver-auth. Then the python script similar to the make-intercept script
calls ninja passing the input/output pipe.

Future versions of make (4.4) can pass --jobserver-auth=fifo:PATH
to pass a named pipe in which case the ninja recipe will need to be updated
to fetch the branch topic-jobserver-fifo and the intercept script will need a
minor modification.

Signed-off-by: Ferry Toth <ftoth@exalondelft.nl>
---
 meta/recipes-devtools/ninja/files/ninja.py    | 37 +++++++++++++++++++
 .../{ninja_1.10.2.bb => ninja_1.12.0.bb}      | 10 +++--
 2 files changed, 44 insertions(+), 3 deletions(-)
 create mode 100755 meta/recipes-devtools/ninja/files/ninja.py
 rename meta/recipes-devtools/ninja/{ninja_1.10.2.bb => ninja_1.12.0.bb} (67%)

diff --git a/meta/recipes-devtools/ninja/files/ninja.py b/meta/recipes-devtools/ninja/files/ninja.py
new file mode 100755
index 0000000000..d3e1b55f25
--- /dev/null
+++ b/meta/recipes-devtools/ninja/files/ninja.py
@@ -0,0 +1,37 @@
+#!/usr/bin/env python3
+import sys
+import os
+import subprocess
+
+if "BB_MAKEFIFO" in os.environ:
+    fifoname = os.environ["BB_MAKEFIFO"]
+
+    r = os.open(fifoname, os.O_RDONLY|os.O_NONBLOCK)
+    w = os.open(fifoname, os.O_WRONLY)
+    # since python 3.3 handles are no longer inheritable by default
+    os.set_inheritable(w, True)
+    os.close(r)
+    r = os.open(fifoname, os.O_RDONLY)
+    os.set_inheritable(r, True)
+
+    # look for -j n and if seen delete it
+    seen = False
+    Next = False
+    NewArgv = []
+    for i in sys.argv:
+        if Next:
+            Next = False
+        else:
+            if "-j" in i:
+                seen = True
+                Next = True
+            else:
+                NewArgv.append(i)
+
+    if seen:
+        os.environ["MAKEFLAGS"] = "-j --jobserver-auth=" + str(r) + "," + str(w)
+
+NewArgv[0] = "ninja.run"
+
+# even if inheritable still nede to explicitly prevent from being closed
+sys.exit(subprocess.call(NewArgv, shell=False, close_fds=False))
diff --git a/meta/recipes-devtools/ninja/ninja_1.10.2.bb b/meta/recipes-devtools/ninja/ninja_1.12.0.bb
similarity index 67%
rename from meta/recipes-devtools/ninja/ninja_1.10.2.bb
rename to meta/recipes-devtools/ninja/ninja_1.12.0.bb
index 7270321d6e..9b32f4c06c 100644
--- a/meta/recipes-devtools/ninja/ninja_1.10.2.bb
+++ b/meta/recipes-devtools/ninja/ninja_1.12.0.bb
@@ -6,9 +6,12 @@ LIC_FILES_CHKSUM = "file://COPYING;md5=a81586a64ad4e476c791cda7e2f2c52e"
 
 DEPENDS = "re2c-native ninja-native"
 
-SRCREV = "e72d1d581c945c158ed68d9bc48911063022a2c6"
+SRCREV = "9fc01d91dedaa3a5584844427ea0fa669fbce59a"
 
-SRC_URI = "git://github.com/ninja-build/ninja.git;branch=release;protocol=https"
+SRC_URI = "git://github.com/stefanb2/ninja.git;branch=topic-tokenpool-master;protocol=https \
+	file://ninja.py \
+	"
+SRC_URI[sha256sum] = "5702446cd3919e73b39e4de8903c087025f068edd11825f271d8b7fa026d71b0"
 UPSTREAM_CHECK_GITTAGREGEX = "v(?P<pver>.*)"
 
 S = "${WORKDIR}/git"
@@ -25,7 +28,8 @@ do_compile() {
 }
 
 do_install() {
-	install -D -m 0755  ${S}/ninja ${D}${bindir}/ninja
+	install -D -m 0755  ${WORKDIR}/ninja.py ${D}${bindir}/ninja
+	install -D -m 0755  ${S}/ninja ${D}${bindir}/ninja.run
 }
 
 BBCLASSEXTEND = "native nativesdk"
-- 
2.37.2

